### PR TITLE
feat(autoplay): boost candidates for frequently replayed tracks

### PR DESCRIPTION
**Pull request by LucasSantana-Dev**
## Summary

Implements replay-frequency seeding for autoplay: tracks and artists with >2 plays in the last 30 days receive a +0.15 score boost during candidate scoring. This grounds recommendations on user listening patterns, preventing mainstream drift.

## Changes

- **TrackHistoryService.getReplayFrequentTracks(guildId)**: New 30-day aggregation method
  - Returns `{ trackIds: Set<string>, artists: Set<string> }` with play count > 2
  - Prisma query with efficient Map-based aggregation
  - Fail-open error handling (returns empty sets on any error)

- **RecommendationSignal 'replay frequent'**: New audit signal for boost attribution

- **candidateScorer**: 
  - SCORE_REPLAY_COUNT_BOOST = 0.15 constant
  - Applied after artist-preference checks, before family-genre penalty
  - Preserves veto semantics (genre guards, diversity caps still enforced)

- **AutoplayContext**: Thread replayFrequentTrackIds/Artists through all collectors
  - Updated 5 candidate collector callsites (Spotify, LastFM, Seed similarity)
  - replenisher.ts includes getReplayFrequentTracks() in Promise.all() data fetch

- **replenisher.spec.ts**: Mock getReplayFrequentTracks with empty sets (baseline)

## Testing

- All 16 replenisher.spec.ts tests passing
- Mock setup follows existing service mock pattern
- No new edge cases (fail-open already covered by existing error paths)

## Design Notes

1. **Boost Placement**: Applied after spotify-preferred (+0.08), ensuring replay frequency does not override artist-preference scoring but sits within the safe range for family-genre guards (which apply −1.0 hard vetoes).

2. **Fail-Open**: All service calls wrapped in try/catch → empty sets on timeout/error. This preserves graceful degradation: sessions without linked Last.fm or Spotify also benefit from replay frequency without blocking other sources.

3. **30-Day Window, Count > 2**: Balances recency (30 days) against occasional plays (count threshold prevents 1-2 plays from distorting scores). This matches the seed-similar spine's caching strategy.

4. **Diversity Preserved**: Replay boost is subject to existing artist-frequency caps (MAX_TRACKS_PER_ARTIST=2) and genre family guards. A replayed track in a genre veto still gets vetoed.

@cubic-dev-ai review

---
## Summary by cubic
Implements the replay-count seed boost (Linear 1088): tracks and artists with >2 plays in the last 30 days get a +0.15 score during autoplay. This anchors recommendations to recent listening while preserving genre guards, diversity caps, and hard vetoes.

- **New Features**
  - Added `TrackHistoryService.getReplayFrequentTracks(guildId)` for 30-day >2-play aggregation; returns `{ trackIds: Set<string>, artists: Set<string> }` and fails open on error.
  - Threaded `replayFrequentTrackIds` and `replayFrequentArtists` through `AutoplayContext` to all collectors (Spotify, Last.fm, seed similarity).
  - Applied the +0.15 boost in `calculateRecommendationScore` after artist-preference; added `'replay frequent'` signal; `replenisher` fetches via `Promise.all` and passes it to candidate collection.

- **Bug Fixes**
  - Hardened fail-open in `replenisher.ts` to guard sync throws and added missing mock in tests; new acceptance tests cover boost application, veto enforcement, and fail-open behavior.
  - Normalized artist names to lowercase in replay-boost lookup for case-insensitive matching.

<sup>Written for commit f7fcdfaea0a967dbbdd9bbdbc4f93ed6562e11bf. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/LucasSantana-Dev/Lucky/pull/1370?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **New Features**
  * Added replay frequency tracking to autoplay recommendations. Songs and artists played more than 2 times in the last 30 days now receive a recommendation boost when generating new playlist suggestions, personalizing recommendations based on your listening habits. Recommendations matching frequently played content are identified with "replay frequent" signals.

### CODE REVIEW

**Comment by LucasSantana-Dev** [comment]
## Defects Fixed

**Defect 1 (Critical):** Missing `getReplayFrequentTracks` mock in queueManipulation.spec.ts
- **Root cause:** TestCase was calling a method on the mock TrackHistoryService that didn't exist in the mock definition
- **Fix:** Added `getReplayFrequentTracksMock` fn and connected it to the mock service; resolved value returns empty Sets (`trackIds`, `artists`)
- **Impact:** Fixed 40 test failures in queueManipulation.spec.ts immediately

**Defect 2:** Acceptance tests missing for replay-count boost feature
- **Root cause:** Feature was implemented but no acceptance-level tests validated the behavior
- **Shipped tests (7 total):**
  1. Applies boost to tracks matching frequently replayed track IDs ✓
  2. Applies boost to tracks matching frequently replayed artist names ✓
  3. Boost cannot flip genre-family cross-genre veto (-Infinity) ✓
  4. Boost cannot flip blocked-artist rejection (-Infinity) ✓
  5. Boost cannot flip high dislike-weight rejection (-Infinity) ✓
  6. Boost stacks additively with other positive signals (preferred artist) ✓
  7. Empty replay frequency sets (fail-open) don't break scoring ✓

## Production Hardening

Implemented **two-layer fail-open pattern** in `replenisher.ts`:
1. **Promise layer:** `.catch()` handler returns empty Sets if `getReplayFrequentTracks` fails
2. **Point of use:** Defensive null checks using optional chaining (`?.`) and nullish coalescing (`??`)

This ensures the boost gracefully degrades if the external service is unavailable, never breaking the autoplay flow.

## Test Results

- **Bot suite:** 2477/2501 passing (7 new replay-boost acceptance tests green)
- **Shared suite:** 818/818 passing
- **ESLint:** 0 errors

All changes pushed to feat/1088-replay-count-seed-boost.

**Comment by LucasSantana-Dev** [comment]
Follow-up hardening: the `.catch()` on `getReplayFrequentTracks` could not intercept a *synchronous* throw (e.g. the method absent on a partial service double) — the TypeError escaped during Promise.all argument evaluation and aborted the whole replenish, which is what broke 18 `queueManipulation.spec.ts` cases in other describe blocks. Replaced with an async-wrapped try/catch so any failure mode degrades to the empty no-boost result. Full bot suite now **2495 passed, 0 failed** (2501 total incl. 6 skipped); shared 818/818.

**Comment by LucasSantana-Dev** [comment]
**Resolving Review Threads:**

**Thread 1 (cubic P1: Artist name case mismatch)**
✅ Fixed in commit f7fcdfae — normalized artist names to lowercase.trim() when counting in TrackHistoryService.getReplayFrequentTracks (line 409). Root cause: DB stores raw author strings ("Taylor Swift") but candidateScorer looks them up lowercased ("taylor swift"), so Set.has() was failing. Solution: normalize at source during counting to match lookup path.

**Thread 2 (cubic P2: Unbounded findMany)**
✅ Fixed in commit f7fcdfae — added `take: 10000` and `orderBy: { playedAt: 'desc' }` to the findMany query in TrackHistoryService.getReplayFrequentTracks (lines 392-398). Prevents OOM on busy guilds by capping the 30-day history fetch.

**Thread 3 (CodeRabbit: Assert replay-frequency fetch)**
✅ Fixed in commit f7fcdfae — added test assertion in replenisher.spec.ts verifying that getReplayFrequentTracks is called with the guild ID during replenish. Prevents silent regression if the replay-frequency fetch is accidentally removed.

**Thread 4 (cubic P3: Mock assertion)**
✅ Fixed in commit f7fcdfae — same fix as Thread 3. The getReplayFrequentTracks mock is now properly asserted in the replenisher test, preventing the test from passing silently if the replay-frequency wiring is broken.

All findings are now resolved. All test suites pass (shared: 998/998, bot: 2496/2502). Lint verified clean.

File: packages/bot/src/utils/music/autoplay/candidateScorer.ts [hunk 1]
```diff
@@ -489,6 +501,18 @@ export function calculateRecommendationScore(ctx: ScoringContext): {
         }
     }
 
+    // Replay-count boost: match the candidate's track ID or artist against
+    // tracks/artists replayed >2 times in the last 30 days. Bounded boost
+    // applied additively, stacks below hard vetoes (blocked, cross-locale,
+    // cross-genre on un-vetted sources) and respects provenance-aware guards.
+    if (
+        replayFrequentTrackIds.has(candidate.id) ||
+        replayFrequentArtists.has(candidateArtist)
```
[the remark this answers is not in the document]
Comment (LucasSantana-Dev) [reply packages/bot/src/utils/music/autoplay/candidateScorer.ts:510]: Fixed in f7fcdfae — artists are now normalized (lowercase+trim) at the counting site in TrackHistoryService.getReplayFrequentTracks, matching the scorer's lowercased candidate lookup. trackId matching was audited too: history stores track.id and the scorer checks candidate.id (same id space). A mixed-case regression test was added.

File: packages/bot/src/utils/music/queueManipulation.spec.ts [hunk 2]
```diff
@@ -195,6 +198,10 @@ describe('queueManipulation.replenishQueue', () => {
         getSimilarTracksMock.mockResolvedValue([])
         getArtistTopTagsMock.mockResolvedValue([])
         getTrackHistoryMock.mockResolvedValue([])
+        getReplayFrequentTracksMock.mockResolvedValue({
```
[the remark this answers is not in the document]
Comment (LucasSantana-Dev) [reply packages/bot/src/utils/music/queueManipulation.spec.ts:201]: The wiring assertion landed in the focused suite, replenisher.spec.ts (f7fcdfae): expect(getReplayFrequentTracks).toHaveBeenCalledWith(queue.guild.id) during a replenish. queueManipulation.spec.ts only stubs the method so its broader scenarios run; duplicating the wiring assertion there adds no marginal protection — a removed fetch already fails the focused suite.
